ykcompile/src/lib.rs

```diff
@@ -114,14 +112,25 @@ pub struct RegAndOffset {
     offs: i32,
 }
 
+/// Describes the location of the pointer in Location::Indirect.
+#[derive(Debug, Clone, PartialEq)]
+pub enum IndirectLoc {
+    /// There's a pointer in this register.
+    Register(u8),
+    /// There's a pointer in memory somewhere.
+    Mem(RegAndOffset),
+}
+
 #[derive(Debug, Clone, PartialEq)]
 pub enum Location {
     /// A value in a register.
     Register(u8),
     /// A statically known memory location relative to a register.
     Mem(RegAndOffset),
-    /// A runtime memory location stored in a register.
-    Addr(u8),
+    /// A location that contains a pointer to some underlying storage.
+    Indirect { ptr: IndirectLoc, offs: i32 },
```

Comment:
Can offsets be negative?

Reply:
In general pointer offsets can be negative. For example `Rc` stores the reference counters at a negative offset from the stored pointer. I don't know if that can leak into here though.

Reply:
For indirect locations, in theory, yes. In practise, not sure.

Reply:
I think it might be worth thinking about this, because if we can't get negative offsets here, a `u32` (or `usize`?!) will save us a few future bugs. [As an aside, I suspect that `offs` / `off` should be `isize`?]

Reply:
In light of @bjorn3's comment, since any kind of smart pointer, including `Rc`, will use indirect locations to realise runtime pointers, I think this should stay signed.

Reply:
OK, then I think it should be `isize`?

Reply:
Hrm, maybe. The reason it's an i32 is that the ISA doesn't allow such big/small offsets.

```
error[E0308]: mismatched types
 --> src/main.rs:7:27
  |
7 |         ; mov rax, [rsp + offs]
  |                           ^^^^ expected `i32`, found `isize`
  |
```

So using an i32 reduces the amount of annoying casting needed, but then again, `Location` ought to be platform agnostic. I don't know if other chips allow such large/small offsets.

Reply:
Another way of fudging is to make a `type` `OffT` which is an alias for `i32` on x86, but which we could realias on other platforms. Though `type` in Rust is rather underpowered, so it can give an illusion of type safety which it doesn't actually possess. There are probably no perfect answers here.

Reply:
It seems a decent compromise. One I've considered before actually.

At least a type alias has a name which can convey meaning, even if it isn't type safe.

Reply:
OK, let's try it?